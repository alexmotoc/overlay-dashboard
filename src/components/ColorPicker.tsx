import * as React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogTitle from '@material-ui/core/DialogTitle';
import PaletteIcon from '@material-ui/icons/Palette';
import { SketchPicker, ColorResult } from 'react-color';

type ColorPickerProps = {
    name: string;
    colorChangeCallback(color: string): void;
};

export const ColorPicker: React.FunctionComponent<ColorPickerProps> = (props: ColorPickerProps) => {
    const [color, setColor] = React.useState('');
    const [isColorPickerOpen, setIsColorPickerOpen] = React.useState(false);

    const handleColorPickerClick = () => {
        setIsColorPickerOpen(true);
    };
    
    const handleCancel = () => {
        setIsColorPickerOpen(false);
    };

    const handleConfirmColor = () => {
        setIsColorPickerOpen(false);
        props.colorChangeCallback(color);
    };

    const handleColorChange = (color: ColorResult) => {
        setColor(color.hex);
        props.colorChangeCallback(color.hex);
    };

    return (
        <React.Fragment>
            <Button
                onClick={handleColorPickerClick}>
                <PaletteIcon color='primary' />
            </Button>
            <Dialog
                open={isColorPickerOpen} >
                <DialogTitle>{props.name} Colour</DialogTitle>
                <DialogContent dividers>
                    <SketchPicker 
                        color={color}
                        onChangeComplete={handleColorChange}/>
                </DialogContent>
                <DialogActions>
                    <Button autoFocus onClick={handleCancel} color="primary">
                    Cancel
                    </Button>
                    <Button onClick={handleConfirmColor} color="primary">
                    OK
                    </Button>
                </DialogActions>
            </Dialog>
        </React.Fragment>

    );
};
