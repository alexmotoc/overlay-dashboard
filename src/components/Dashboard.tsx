import * as React from 'react';
import * as _ from "lodash";
import axios from 'axios';
import moment from 'moment';
import Grid from '@material-ui/core/Grid';
import PeopleIcon from '@material-ui/icons/People';
import TimelapseIcon from '@material-ui/icons/Timelapse';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import TwitchClient from 'twitch';
import ChatClient from 'twitch-chat-client';

const useStyles = makeStyles({
    gridContainer: {
        marginTop: 50
    },
    streamInfoContainer: {
        marginLeft: 50
    },
    chatContainer: {
        marginRight: 50
    },
    thumbnail: {
        marginTop: 15,
        width: '100%'
    },
    infoContainer: {
        display: 'flex',
        justifyContent: 'space-between',
        marginTop: 10
    },
    statContainer: {
        display: 'flex',
        alignItems: 'center'
    },
    countMargin: {
        marginLeft: 15
    },
    chat: {
        maxHeight: '80%',
        overflow: 'auto',
        marginLeft: 50,
        marginTop: 25
    },
    message: {
        display: 'flex',
        flexDirection: 'row',
        marginTop: 5
    },
    user: {
        fontWeight: 'bold'
    },
    offline: {
        marginTop: '25%'
    }
});

type Chat = {
    time: string;
    user: string;
    message: string;
};

export const Dashboard: React.FunctionComponent<{}> = () => {
    const classes = useStyles();

    const [title, setTitle] = React.useState<string>('');
    const [viewerCount, setViewerCount] = React.useState<number>(0);
    const [duration, setDuration] = React.useState<string>('');
    const [thumbnail, setThumbnail] = React.useState<string>('');
    const [chat, setChat] = React.useState<Chat[]>([]);
    const [isOnline, setIsOnline] = React.useState<boolean>(false);

    // Connect to the Twitch channel's chat
    (async () => {
        const clientId = '0qmxdyuchkdcpkfktmq2t47z06eng1';
        const clientSecret = '7b00yip56b78z0kiwt9bvduia5oinn';

        let response = await axios.get('https://tungsten.alexlogan.co.uk/user/b0960c68-af68-4e5b-8447-1150878998c1/');
        const tokenData = JSON.parse(response.data.tokens);

        // Use automatically-refreshed tokens for the connection
        const twitchClient = TwitchClient.withCredentials(clientId, tokenData.twitch.chat.accessToken, undefined, {
            clientSecret,
            refreshToken: tokenData.twitch.chat.refreshToken,
            expiry: tokenData.twitch.chat.expiryTimestamp === null ? null : new Date(tokenData.twitch.chat.expiryTimestamp),
            onRefresh: async ({ accessToken, refreshToken, expiryDate }) => {
                const newTokenData = {
                    accessToken,
                    refreshToken,
                    expiryTimestamp: expiryDate === null ? null : expiryDate.getTime()
                };
                tokenData.twitch.chat = newTokenData;
                await axios.put('https://tungsten.alexlogan.co.uk/user/b0960c68-af68-4e5b-8447-1150878998c1/', {'tokens': JSON.stringify(tokenData)});
            }
        });

        // Get channel name
        let user: string = '';
        response = await axios.get('https://api.twitch.tv/helix/users', { headers: {"Authorization" : `Bearer ${tokenData.twitch.authentication}`}});
        if (response.data.data) {
            user = response.data.data[0].login;
        }
    
        const chatClient = await ChatClient.forTwitchClient(twitchClient, {channels: [user]});
        await chatClient.connect();
    
        // Subscribe to the new message event
        chatClient.onPrivmsg((channel, user, message) => {
            const newChat: Chat[] = chat;
            const time: string = `${moment().hour()}:${moment().minute()}`;
            if (newChat.length === 0 || (newChat.length > 0 && !_.isEqual(newChat[newChat.length - 1], {time: time, user: user, message: message}))) {
                newChat.push({time: time, user: user, message: message});
                setChat([...newChat]);
            }
        });
    })();

    const fetchStreamData = (stream: string) => {
        // Get stream related information
        axios.get(`https://api.twitch.tv/helix/streams?user_login=${stream}`, {headers: {'Client-ID': '0qmxdyuchkdcpkfktmq2t47z06eng1'}}).then(response => {
            const data = response.data.data[0];

            // Check if stream is online
            if (data) {
                setIsOnline(true);
                const diff = moment.duration(moment().diff(moment(data.started_at)));
            
                if (diff.hours() === 0) {
                    setDuration(diff.minutes() + 'm');
                } else {
                    setDuration(diff.hours() + 'h:' + diff.minutes() + 'm');
                }
                
                setTitle(data.title);
                setThumbnail(data.thumbnail_url.replace('{width}x{height}', '1920x1080'));
                setViewerCount(data.viewer_count);
            } else {
                setIsOnline(false);
            }
        });
    };

    React.useEffect(() => {
        // Get the OAuth token of the logged in account
        axios.get('https://tungsten.alexlogan.co.uk/user/b0960c68-af68-4e5b-8447-1150878998c1/').then(response => {
            const tokenData = JSON.parse(response.data.tokens);

            axios.get('https://api.twitch.tv/helix/users', { headers: {"Authorization" : `Bearer ${tokenData.twitch.authentication}`}}).then(response => {
                if (response.data.data) {
                    const stream: string = response.data.data[0].login;

                    // Update statistics every minute
                    fetchStreamData(stream);    
                    const interval = setInterval(() => {
                        fetchStreamData(stream);
                    }, 60000);
                    return () => clearInterval(interval);
                }
            });
        });
    }, []);

    return (
        <React.Fragment>
            <Typography variant='h3' align='center'>
                Stream Information
            </Typography>
            {isOnline &&
                <Grid container className={classes.gridContainer} spacing={2}>
                    <Grid item xs className={classes.streamInfoContainer}>
                        <Typography variant="h4" align="center">
                            {title}
                        </Typography> 
                        <img className={classes.thumbnail} src={thumbnail} alt="Stream Thumbnail"/> 
                        <div className={classes.infoContainer}>
                            <div className={classes.statContainer}> 
                                <PeopleIcon fontSize='large'/>
                                <Typography className={classes.countMargin}>
                                    {viewerCount}
                                </Typography>
                            </div>   
                            <div className={classes.statContainer}>
                                <TimelapseIcon fontSize='large'/>
                                <Typography className={classes.countMargin}>
                                    {duration}
                                </Typography>  
                            </div>  
                        </div>                 
                    </Grid>
                    <Grid item xs className={classes.chatContainer}>
                        <Typography variant="h4" align="center">
                            Chat
                        </Typography>
                        <div className={classes.chat}>
                            {chat.reverse().map((el, idx) => (
                                <div key={idx} className={classes.message}>
                                    <Typography className={classes.user}>
                                    [{el.time}] {el.user}:
                                    </Typography>
                                    <Typography>
                                        {el.message}
                                    </Typography>
                                </div>
                            ))}
                        </div>
                    </Grid>
                </Grid>
            }
            {!isOnline &&
                <Typography className={classes.offline} variant='h4' align='center'>
                    Stream is currently offline
                </Typography>
            }
        </React.Fragment>
    );
}