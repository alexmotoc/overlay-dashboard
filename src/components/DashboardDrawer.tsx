import * as React from 'react';
import Drawer from '@material-ui/core/Drawer';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import ListItemText from '@material-ui/core/ListItemText';
import BuildIcon from '@material-ui/icons/Build';
import HomeIcon from '@material-ui/icons/Home';
import SettingsIcon from '@material-ui/icons/Settings';
import { Theme } from '@material-ui/core';
import {createStyles, makeStyles } from '@material-ui/styles';

const drawerWidth: number = 240;

type Option = {
    name: string;
    link: string;
    icon: JSX.Element;
};

const options: Option[] = [
    {'name': 'Dashboard', 'link': '/dashboard', 'icon': <HomeIcon/>},
    {'name': 'Stream Settings', 'link': '/settings', 'icon': <SettingsIcon/>},
    {'name': 'Overlay Builder', 'link': '/overlay-builder', 'icon': <BuildIcon/>}
];

const useStyles = makeStyles((theme: Theme) => 
    createStyles({
        drawer: {
            [theme.breakpoints.up('sm')]: {
              width: 240,
              flexShrink: 0,
            },        
          },
        drawerPaper: {
            width: drawerWidth,
            backgroundColor: theme.palette.primary.main
        }
    })
);

export const DashboardDrawer: React.FunctionComponent<{}> = () => {
    const classes = useStyles();

    const drawer: React.ReactFragment= (
        <List>
            {options.map((option, _) => (
            <ListItem button component='a' key={option.name} href={option.link}>
                <ListItemIcon>
                    {option.icon}
                </ListItemIcon>
                <ListItemText primary={option.name} />
            </ListItem>
            ))}
        </List>
    );

    return (
        <nav className={classes.drawer}>
            <Drawer
            classes={{
            paper: classes.drawerPaper,
            }}
            variant="permanent"
            open
        >
            {drawer}
        </Drawer>
        </nav>
    );
}