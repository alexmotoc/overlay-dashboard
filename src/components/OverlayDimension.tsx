import * as React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Input from '@material-ui/core/Input';
import Typography from '@material-ui/core/Typography';
import Slider from '@material-ui/core/Slider';
import HeightIcon from '@material-ui/icons/Height';

const useStyles = makeStyles({
    root: {
      width: 250,
    },
    input: {
      width: 42,
    },
    rotate90: {
        transform: 'rotate(90deg)'
    },
  });

type OverlayDimensionProps = {
    name: string;
};

export const OverlayDimension: React.FunctionComponent<OverlayDimensionProps> = (props: OverlayDimensionProps) => {
    const classes = useStyles();
    const [value, setValue] = React.useState(0);

    const handleSliderChange = (_: React.ChangeEvent<{}>, newValue: number | number[]) => {
        setValue(Number(newValue));
    };

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setValue(Number(event.target.value));
    };

    const handleBlur = () => {
        if (value < 0) {
          setValue(0);
        } else if (value > 100) {
          setValue(100);
        }
    };

    return (
        <div className={classes.root}>
            <Typography id="input-slider" gutterBottom>
            {props.name} (%)
            </Typography>
            <Grid container spacing={2} alignItems="center">
                <Grid item>
                    {props.name === 'Height' ? (
                        <HeightIcon />
                    ) : (
                        <HeightIcon className={classes.rotate90} />
                    )
                    }
                </Grid>
                <Grid item xs>
                    <Slider
                        value={value}
                        onChange={handleSliderChange}
                        min={0}
                        max={100}
                        aria-labelledby="input-slider"
                    />
                </Grid>
                <Grid item>
                    <Input
                        className={classes.input}
                        value={value}
                        margin="dense"
                        onChange={handleInputChange}
                        onBlur={handleBlur}
                        inputProps={{
                        step: 5,
                        min: 0,
                        max: 100,
                        type: 'number',
                        'aria-labelledby': 'input-slider',
                        }}
                    />
                </Grid>
            </Grid>
        </div>
    );
}