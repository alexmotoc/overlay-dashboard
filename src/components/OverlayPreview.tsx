import * as React from 'react';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles({
    overlay: {
      width: (props: OverlayPreviewProps) => props.width,
      height: (props: OverlayPreviewProps) => props.height,
      backgroundColor: 'black',
      textAlign: 'center'
    },
    text: {
        display: 'block',
        position: 'relative',
        top: '50%',
        transform: 'translateY(-50%)',
        color: (props: OverlayPreviewProps) => props.textColor
    }
});

type OverlayPreviewProps = {
    width: number;
    height: number;
    text: string;
    textColor: string;
};

export const OverlayPreview: React.FunctionComponent<OverlayPreviewProps> = (props: OverlayPreviewProps) => {
    const classes = useStyles(props);
    
    return (
        <div className={classes.overlay}> 
            <span className={classes.text}>{props.text}</span>         
        </div>
    );
}