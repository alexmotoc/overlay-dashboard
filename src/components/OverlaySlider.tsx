import * as React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Input from '@material-ui/core/Input';
import Typography from '@material-ui/core/Typography';
import Slider from '@material-ui/core/Slider';
import HeightIcon from '@material-ui/icons/Height';
import LocationOnIcon from '@material-ui/icons/LocationOn';

const useStyles = makeStyles({
    root: {
      width: 250,
    },
    input: {
      width: 42,
    },
    rotate90: {
        transform: 'rotate(90deg)'
    },
});

type OverlaySliderProps = {
    type: string;
    name: string;
    min: number;
    max: number;
    onValueChange(attribute: string, value: number): void;
};

export const OverlaySlider: React.FunctionComponent<OverlaySliderProps> = (props: OverlaySliderProps) => {
    const classes = useStyles();
    const [value, setValue] = React.useState(0);

    const handleSliderChange = (_: React.ChangeEvent<{}>, newValue: number | number[]) => {
        setValue(Number(newValue));
        props.onValueChange(props.name, Number(newValue));
    };

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setValue(Number(event.target.value));
        props.onValueChange(props.name, Number(event.target.value));
    };

    const handleBlur = () => {
        if (value < 0) {
          setValue(0);
        } else if (value > 100) {
          setValue(100);
        }
    };

    return (
        <div className={classes.root}>
            <Typography id="input-slider" gutterBottom>
            {props.name} {props.type === 'size' ? '(%)' : '(px)'} 
            </Typography>
            <Grid container spacing={2} alignItems="center">
                <Grid item>
                    {props.type === 'size' ? (
                        props.name === 'Height' ? (
                            <HeightIcon />
                        ) : (
                            <HeightIcon className={classes.rotate90} />
                        )
                    ) : (
                        <LocationOnIcon />
                    )}
                </Grid>
                <Grid item xs>
                    <Slider
                        value={value}
                        onChange={handleSliderChange}
                        min={props.min}
                        max={props.max}
                        aria-labelledby="input-slider"
                    />
                </Grid>
                <Grid item>
                    <Input
                        className={classes.input}
                        value={value}
                        margin="dense"
                        onChange={handleInputChange}
                        onBlur={handleBlur}
                        inputProps={{
                        step: 5,
                        min: props.min,
                        max: props.max,
                        type: 'number',
                        'aria-labelledby': 'input-slider',
                        }}
                    />
                </Grid>
            </Grid>
        </div>
    );
}