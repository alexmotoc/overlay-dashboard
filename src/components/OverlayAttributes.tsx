import * as React from 'react';
import IconButton from '@material-ui/core/IconButton';
import TextField from '@material-ui/core/TextField';
import DeleteIcon from '@material-ui/icons/Delete';
import { ColorPicker } from './ColorPicker';
import { OverlaySlider } from './OverlaySlider';
import { TextAlignment, TextStyle, TextFormats } from './TextStyle';
import { Typography } from '@material-ui/core';
import { streamHeight, streamWidth } from './OverlayBuilder';
import red from '@material-ui/core/colors/red';
import { makeStyles } from '@material-ui/core/styles';
import { RGBColor } from 'react-color';

export type Size = {
    width: number;
    height: number;
};

export type Position = {
    horizontal: number;
    vertical: number;
};

export type Text = {
    content: string;
    alignment: TextAlignment;
    colour: RGBColor;
    format: TextFormats[];
};

export type Overlay = {
    name?: string;
    size: Size;
    position: Position;
    colour: RGBColor;
    text: Text;
};

export type Template = {
    name?: string;
    overlays: Overlay[];
}

type OverlayAttributesProps = {
    overlay: Overlay;
    onChange(overlay: Overlay): void;
    onDeleteOverlay(): void;
    showDelete: boolean;
};

export const white: RGBColor = {r: 255, g: 255, b: 255, a: 1.0};
export const black: RGBColor = {r: 0, g: 0, b: 0, a: 1.0};

export const defaultOverlay: Overlay = {
    name: 'Overlay',
    size: {
        width: 0,
        height: 0
    },
    position: {
        horizontal: 0,
        vertical: 0
    },
    colour: black,
    text: {
        content: '',
        alignment: 'center' as TextAlignment,
        format: ['' as TextFormats],
        colour: white
    }
};

const useStyles = makeStyles({
    form: {
        marginLeft: 50
    },
    formDisplay: {
        display: 'flex',
        flexDirection: 'column'
    },
    previewArea: {
        textAlign: 'center',
        marginRight: 50
    },
    headerText: {
        marginTop: 25,
        fontWeight: 'bold'
    },
    element: {
        marginTop: 15,
        width: '80%'
    },
    textStyle: {
        display: 'flex'
    },
    extendedIcon: {
        marginRight: 10,
    },
    saveButton: {
        marginTop: 15
    },
    overlayName: {
        fontSize: '3em'
    },
    overlayTabs: {
        display: 'flex',
        flexDirection: 'row',
        maxWidth: 500
    },
    overlayNameContainer: {
        display: 'flex',
        alignItems: 'center'
    },
    deleteItem: {
        marginRight: 10,
        color: red[500]
    }
});

export const OverlayAttributes: React.FunctionComponent<OverlayAttributesProps> = (props: OverlayAttributesProps) => { 
    const classes = useStyles();
    const overlay: Overlay = props.overlay;

    const [overlayName, setOverlayName] = React.useState(overlay.name);
    const [width, setWidth] = React.useState(overlay.size.width);
    const [height, setHeight] = React.useState(overlay.size.height);
    const [horizontal, setHorizontal] = React.useState(overlay.position.horizontal);
    const [vertical, setVertical] = React.useState(overlay.position.vertical);
    const [text, setText] = React.useState(overlay.text.content);
    const [textAlignment, setTextAlignment] = React.useState(overlay.text.alignment);
    const [textFormats, setTextFormats] = React.useState(overlay.text.format);
    const [textColor, setTextColor] = React.useState(overlay.text.colour);
    const [overlayColor, setOverlayColor] = React.useState(overlay.colour);

    const getCurrentOverlay = () => {
        return {
            name: overlayName,
            size: {
                width: width,
                height: height
            },
            position: {
                horizontal: horizontal,
                vertical: vertical
            },
            colour: overlayColor,
            text: {
                content: text,
                alignment: textAlignment,
                format: textFormats,
                colour: textColor
            }
        };
    };

    const handleValueChange = (attribute: string, value: number) => {
        switch (attribute.toLowerCase()) {
            case 'width':
                setWidth(value);
                break;
            case 'height':
                setHeight(value);
                break;
            case 'horizontal':
                setHorizontal(value);
                break;
            case 'vertical':
                setVertical(value);
                break;
        }
        props.onChange(getCurrentOverlay());
    };

    const handleTextChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setText(event.target.value);
        const currentOverlay: Overlay = getCurrentOverlay();
        currentOverlay.text.content = event.target.value;
        props.onChange(currentOverlay);
    };

    const handleOverlayNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setOverlayName(event.target.value);
        const currentOverlay: Overlay = getCurrentOverlay();
        currentOverlay.name = event.target.value;
        props.onChange(currentOverlay);
    };

    const handleColorChange = (color: RGBColor) => {
        setOverlayColor(color);
        const currentOverlay: Overlay = getCurrentOverlay();
        currentOverlay.colour = color;
        props.onChange(currentOverlay);
    };

    const handleTextAlignmentChange = (newAlignment: TextAlignment) => {
        setTextAlignment(newAlignment);
        const currentOverlay: Overlay = getCurrentOverlay();
        currentOverlay.text.alignment = newAlignment;
        props.onChange(currentOverlay);
    };

    const handleTextColorChange = (color: RGBColor) => {
        setTextColor(color);
        const currentOverlay: Overlay = getCurrentOverlay();
        currentOverlay.text.colour = color;
        props.onChange(currentOverlay);
    };

    const handleTextFormatsChange = (newFormats: TextFormats[]) => {
        setTextFormats(newFormats);
        const currentOverlay: Overlay = getCurrentOverlay();
        currentOverlay.text.format = newFormats;
        props.onChange(currentOverlay);
    };

    return (
        <React.Fragment>
            <div className={`${classes.element} ${classes.overlayNameContainer}`}>
                {props.showDelete && <IconButton className={classes.deleteItem} aria-label="delete" onClick={props.onDeleteOverlay}>
                    <DeleteIcon />
                </IconButton>}
                <TextField 
                    value={overlayName}
                    InputProps={{
                        disableUnderline: true,
                        classes: {
                        input: classes.overlayName,
                        },
                    }}
                    onChange={handleOverlayNameChange}/>
            </div>
            <Typography variant='h5' className={classes.headerText}>
                Size
            </Typography>
            <OverlaySlider 
                className={classes.element}
                type='size' 
                name='Height' 
                min={0}
                max={100}
                value={height}
                onValueChange={handleValueChange}/>
            <OverlaySlider 
                className={classes.element}
                type='size' 
                name='Width' 
                min={0}
                max={100}
                value={width}
                onValueChange={handleValueChange}/>
            <Typography variant='h5' className={classes.headerText}>
                Position
            </Typography> 
            <OverlaySlider 
                className={classes.element}
                type='position' 
                name='Horizontal' 
                min={0}
                max={streamWidth}
                value={horizontal}
                onValueChange={handleValueChange}/>
            <OverlaySlider 
                className={classes.element}
                type='position' 
                name='Vertical' 
                min={0}
                max={streamHeight}
                value={vertical}
                onValueChange={handleValueChange}/>
            <div>
                <Typography variant='h5' className={classes.headerText}>
                    Colour
                </Typography> 
                <ColorPicker
                    name='Overlay'
                    color={overlayColor}
                    colorChangeCallback={handleColorChange}>
                </ColorPicker> 
            </div>                       
            <Typography variant='h5' className={classes.headerText}>
                Text
            </Typography>
            <TextField 
                className={classes.element}
                disabled={width === 0 || height === 0}
                variant='outlined'
                value={text}
                onChange={handleTextChange} /> 
            <TextStyle
                className={`${classes.element} ${classes.textStyle}`}
                textAlignmentCallback={handleTextAlignmentChange}
                textColorCallback={handleTextColorChange}
                textFormatsCallback={handleTextFormatsChange}/>
        </React.Fragment>
    )
}