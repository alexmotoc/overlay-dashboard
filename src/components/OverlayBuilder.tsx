import * as React from 'react';
import axios from 'axios';
import AddIcon from '@material-ui/icons/Add';
import Fab from '@material-ui/core/Fab';
import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import SaveIcon from '@material-ui/icons/Save';
import Snackbar from '@material-ui/core/Snackbar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import MuiAlert, { AlertProps } from '@material-ui/lab/Alert';
import grey from '@material-ui/core/colors/grey';
import red from '@material-ui/core/colors/red';
import { makeStyles } from '@material-ui/core/styles';
import { OverlayPreview } from './OverlayPreview';
import { defaultOverlay, Overlay, OverlayAttributes } from './OverlayAttributes';
import { previewWidth, previewHeight } from './OverlayPreview';

export const streamWidth: number = 1920;
export const streamHeight: number = 1080;

const Alert = (props: AlertProps) => {
    return <MuiAlert elevation={6} variant="filled" {...props} />;
}
  
const useStyles = makeStyles({
    form: {
        marginLeft: 50
    },
    formDisplay: {
        display: 'flex',
        flexDirection: 'column'
    },
    previewArea: {
        textAlign: 'center',
        marginRight: 50
    },
    headerText: {
        marginTop: 25,
        fontWeight: 'bold'
    },
    element: {
        marginTop: 15,
        width: '80%'
    },
    textStyle: {
        display: 'flex'
    },
    extendedIcon: {
        marginRight: 10,
    },
    saveButton: {
        marginLeft: 20
    },
    saveContainer: {
        display: 'flex',
        alignItems: 'center',
        marginTop: 15
    },
    overlayName: {
        fontSize: '3em'
    },
    overlayTabs: {
        display: 'flex',
        flexDirection: 'row',
        maxWidth: 500
    },
    overlayNameContainer: {
        display: 'flex',
        alignItems: 'center'
    },
    deleteItem: {
        marginRight: 10,
        color: red[500]
    },
    preview: {
        display: 'inline-block',
        width: previewWidth,
        height: previewHeight,
        marginTop: 50,
        position: 'relative',
        backgroundColor: grey[300],
        overflow: 'hidden'
    },
});

export const OverlayBuilder: React.FunctionComponent<{}> = () => {   
    const classes = useStyles();

    const [isToastOpen, setIsToastOpen] = React.useState<boolean>(false);
    const [overlays, setOverlays] = React.useState<Overlay[]>([defaultOverlay]);
    const [index, setIndex] = React.useState<number>(0);
    const [templateName, setTemplateName] = React.useState<string>('');

    const handleIndexChange = (_: React.ChangeEvent<{}>, newIndex: number) => {
        setIndex(newIndex);
    };

    const handleAddOverlay = () => {
        let newOverlay: Overlay = {...defaultOverlay};
        newOverlay.name = 'Overlay' + overlays.length;
        setOverlays(overlays.concat(newOverlay));
    };

    const handleDeleteOverlay = () => {
        setOverlays(overlays.slice(0, index).concat(overlays.slice(index + 1)));
        setIndex(0);
    };

    const handleOverlayAttributeChange = (overlay: Overlay) => {
        overlays[index] = overlay;
        setOverlays([...overlays]);
    };

    const handleTemplateNameChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setTemplateName(event.target.value);
    };

    const handleSaveOverlay = () => {
        axios.post('http://127.0.0.1:8000/overlays/', {'name': templateName, 'template': JSON.stringify(overlays)}).then(response => {
            setIsToastOpen(true);
        });
    }

    const handleCloseToast = () => {
        setIsToastOpen(false);
    }

    return (
        <React.Fragment>
            <Grid container spacing={2}>
                <Grid item xs className={classes.form}>
                    <div className={classes.formDisplay}>
                        <div className={classes.overlayTabs}>
                            <Tabs
                                value={index}
                                onChange={handleIndexChange}
                                indicatorColor="primary"
                                textColor="primary"
                                variant="scrollable"
                                scrollButtons="off"
                                >
                                {overlays.map((overlay, idx) => (
                                    <Tab label={overlay.name} />
                                ))}
                            </Tabs>
                            <IconButton aria-label="add" color="primary" onClick={handleAddOverlay}>
                                <AddIcon />
                            </IconButton>
                        </div>
                    </div>
                    {overlays.map((overlay, idx) => (
                        idx === index &&
                        <OverlayAttributes
                            overlay={overlay}
                            onChange={handleOverlayAttributeChange}
                            onDeleteOverlay={handleDeleteOverlay}
                            showDelete={overlays.length > 1}
                        />
                    ))}
                    <div className={classes.saveContainer}>
                        <TextField onChange={handleTemplateNameChange} value={templateName} label="Template Name" variant="outlined"/>
                        <Fab className={classes.saveButton} color="primary" variant="extended" onClick={handleSaveOverlay}>
                            <SaveIcon className={classes.extendedIcon}/>
                            Save
                        </Fab>
                    </div>                
                </Grid>
                <Grid item xs className={classes.previewArea}>
                    <Typography variant='h3' align='center'>
                        Template Preview
                    </Typography>
                    <div className={classes.preview}>
                        {overlays.map((overlay, idx) => (
                            <OverlayPreview {...overlay} />
                        ))}
                    </div>
                </Grid>
            </Grid>
            <Snackbar open={isToastOpen} autoHideDuration={1500} onClose={handleCloseToast}>
                <Alert onClose={handleCloseToast} severity="success">
                Overlay was saved successfully!
                </Alert>
            </Snackbar>
        </React.Fragment>
    );
}