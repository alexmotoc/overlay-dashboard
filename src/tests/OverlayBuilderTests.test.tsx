import React from 'react';
import { mount, shallow } from 'enzyme';
import { OverlayBuilder } from '../components/OverlayBuilder';
import { OverlayPreview } from '../components/OverlayPreview';

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useLocation: () => ({
      pathname: "localhost:3000/overlays/"
    })
}));

describe('<OverlaySlider />', () => {

  it('should render correctly', () => {
    const wrapper = mount(<OverlayBuilder />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should display box with 100% width and 100% height', () => {
    const wrapper = mount(<OverlayBuilder />);
    const height = wrapper.find('input').at(2);
    const width = wrapper.find('input').at(4);

    height.simulate('change', { target: { value: '100' } });
    width.simulate('change', { target: { value: '100' } });
    
    const attributes = wrapper.find(OverlayPreview).props();
    expect(attributes.size.height).toEqual(100);
    expect(attributes.size.width).toEqual(100);
  });
});