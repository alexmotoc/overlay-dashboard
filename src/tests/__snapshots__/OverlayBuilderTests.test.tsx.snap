// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OverlaySlider /> should render correctly 1`] = `ReactWrapper {}`;
